### ISSUE TITLE
Does actualvim support Windows OS?

### ISSUE THREAD

**Issue by llinfeng** [lunixbochs/actualvim#32]
It sounds like a good idea to enable Vim-motions in Sublime. Is it compatible on a Windows machine?

If so, could any help with giving a thorough instruction on steps for installation? 

All the best,
- Linfeng

**Comment by xelra** [comment]
It should work with Neovim. Albeit in headless mode, Neovim runs on Windows. But headless is all that actualvim should really need.

**Comment by lunixbochs** [comment]
I'm extremely close to initial neovim release. I need someone to test on Windows. I'm using a single Neovim instance with `--embed`.

**Comment by xelra** [comment]
I can test it on Windows. It would be great if it had the ability to specify a specific, already running Neovim process. Like the `--server` option in Neovim-Qt.

**Comment by lunixbochs** [comment]
I don't really support exposing existing vim buffers/windows yet. I just abuse neovim with a dynamic set of buffers to handle keyboard input. So I'm not sure what connecting to an existing neovim instance would do for you besides spew buffers into it. How would that work for you ideally? Which Sublime features do you want to use? Because right now just about everything in sublime (plugins, search/replace, etc) works besides multiple selection.

Try out the current neovim branch. I mostly want to know if it starts up on Windows. I probably need to add a setting for the Neovim path (though if it has an installer I can default to that path). It should be fairly stable, but it hangs Sublime if you do some chained commands (like gg, 9j, etc).

**Comment by xelra** [comment]
I've cloned the neovim branch into `C:\Program Files\Sublime Text 3\Packages\actualvim`, but when I start up Sublime Text 3, nothing happens.

**Comment by lunixbochs** [comment]
check the st3 console, does it get imported? does it throw errors? I'd expect it to fail to launch nvim.exe.

**Comment by xelra** [comment]
I had to move it to `%APPDATA%\Sublime Text 3\Packages\actualvim`

```
Traceback (most recent call last):
  File "C:\Users\xelra\AppData\Roaming\Sublime Text 3\Packages\actualvim\lib\neovim\msgpack_rpc\event_loop\__init__.py", line 7, in <module>
    from .uv import UvEventLoop
  File "C:\Users\xelra\AppData\Roaming\Sublime Text 3\Packages\actualvim\lib\neovim\msgpack_rpc\event_loop\uv.py", line 5, in <module>
    import pyuv
ImportError: No module named 'pyuv'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\xelra\AppData\Roaming\Sublime Text 3\Packages\actualvim\lib\asyncio\__init__.py", line 13, in <module>
    from . import _overlapped
ImportError: cannot import name _overlapped

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Program Files\Sublime Text 3\sublime_plugin.py", line 109, in reload_plugin
    m = importlib.import_module(modulename)
  File "./python3.3/importlib/__init__.py", line 90, in import_module
  File "<frozen importlib._bootstrap>", line 1584, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1565, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1532, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 584, in _check_name_wrapper
  File "<frozen importlib._bootstrap>", line 1022, in load_module
  File "<frozen importlib._bootstrap>", line 1003, in load_module
  File "<frozen importlib._bootstrap>", line 560, in module_for_loader_wrapper
  File "<frozen importlib._bootstrap>", line 868, in _load_module
  File "<frozen importlib._bootstrap>", line 313, in _call_with_frames_removed
  File "C:\Users\xelra\AppData\Roaming\Sublime Text 3\Packages\actualvim\actual.py", line 4, in <module>
    from .view import ActualVim
  File "C:\Users\xelra\AppData\Roaming\Sublime Text 3\Packages\actualvim\view.py", line 4, in <module>
    from . import neo
  File "C:\Users\xelra\AppData\Roaming\Sublime Text 3\Packages\actualvim\neo.py", line 4, in <module>
    from .lib import neovim
  File "C:\Users\xelra\AppData\Roaming\Sublime Text 3\Packages\actualvim\lib\neovim\__init__.py", line 11, in <module>
    from .msgpack_rpc import (ErrorResponse, child_session, socket_session,
  File "C:\Users\xelra\AppData\Roaming\Sublime Text 3\Packages\actualvim\lib\neovim\msgpack_rpc\__init__.py", line 8, in <module>
    from .event_loop import EventLoop
  File "C:\Users\xelra\AppData\Roaming\Sublime Text 3\Packages\actualvim\lib\neovim\msgpack_rpc\event_loop\__init__.py", line 12, in <module>
    from .asyncio import AsyncioEventLoop
  File "C:\Users\xelra\AppData\Roaming\Sublime Text 3\Packages\actualvim\lib\neovim\msgpack_rpc\event_loop\asyncio.py", line 17, in <module>
    from actualvim.lib import asyncio
  File "C:\Users\xelra\AppData\Roaming\Sublime Text 3\Packages\actualvim\lib\asyncio\__init__.py", line 15, in <module>
    import _overlapped  # Will also be exported.
ImportError: No module named '_overlapped'
reloading plugin actualvim.edit
reloading plugin actualvim.neo
Traceback (most recent call last):
  File "C:\Users\xelra\AppData\Roaming\Sublime Text 3\Packages\actualvim\lib\neovim\msgpack_rpc\event_loop\__init__.py", line 7, in <module>
    from .uv import UvEventLoop
  File "C:\Users\xelra\AppData\Roaming\Sublime Text 3\Packages\actualvim\lib\neovim\msgpack_rpc\event_loop\uv.py", line 5, in <module>
    import pyuv
ImportError: No module named 'pyuv'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\xelra\AppData\Roaming\Sublime Text 3\Packages\actualvim\lib\asyncio\__init__.py", line 13, in <module>
    from . import _overlapped
ImportError: cannot import name _overlapped

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Program Files\Sublime Text 3\sublime_plugin.py", line 109, in reload_plugin
    m = importlib.import_module(modulename)
  File "./python3.3/importlib/__init__.py", line 90, in import_module
  File "<frozen importlib._bootstrap>", line 1584, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1565, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1532, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 584, in _check_name_wrapper
  File "<frozen importlib._bootstrap>", line 1022, in load_module
  File "<frozen importlib._bootstrap>", line 1003, in load_module
  File "<frozen importlib._bootstrap>", line 560, in module_for_loader_wrapper
  File "<frozen importlib._bootstrap>", line 868, in _load_module
  File "<frozen importlib._bootstrap>", line 313, in _call_with_frames_removed
  File "C:\Users\xelra\AppData\Roaming\Sublime Text 3\Packages\actualvim\neo.py", line 4, in <module>
    from .lib import neovim
  File "C:\Users\xelra\AppData\Roaming\Sublime Text 3\Packages\actualvim\lib\neovim\__init__.py", line 11, in <module>
    from .msgpack_rpc import (ErrorResponse, child_session, socket_session,
  File "C:\Users\xelra\AppData\Roaming\Sublime Text 3\Packages\actualvim\lib\neovim\msgpack_rpc\__init__.py", line 8, in <module>
    from .event_loop import EventLoop
  File "C:\Users\xelra\AppData\Roaming\Sublime Text 3\Packages\actualvim\lib\neovim\msgpack_rpc\event_loop\__init__.py", line 12, in <module>
    from .asyncio import AsyncioEventLoop
  File "C:\Users\xelra\AppData\Roaming\Sublime Text 3\Packages\actualvim\lib\neovim\msgpack_rpc\event_loop\asyncio.py", line 17, in <module>
    from actualvim.lib import asyncio
  File "C:\Users\xelra\AppData\Roaming\Sublime Text 3\Packages\actualvim\lib\asyncio\__init__.py", line 15, in <module>
    import _overlapped  # Will also be exported.
ImportError: No module named '_overlapped'
reloading plugin actualvim.view
Traceback (most recent call last):
  File "C:\Users\xelra\AppData\Roaming\Sublime Text 3\Packages\actualvim\lib\neovim\msgpack_rpc\event_loop\__init__.py", line 7, in <module>
    from .uv import UvEventLoop
  File "C:\Users\xelra\AppData\Roaming\Sublime Text 3\Packages\actualvim\lib\neovim\msgpack_rpc\event_loop\uv.py", line 5, in <module>
    import pyuv
ImportError: No module named 'pyuv'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\xelra\AppData\Roaming\Sublime Text 3\Packages\actualvim\lib\asyncio\__init__.py", line 13, in <module>
    from . import _overlapped
ImportError: cannot import name _overlapped

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Program Files\Sublime Text 3\sublime_plugin.py", line 109, in reload_plugin
    m = importlib.import_module(modulename)
  File "./python3.3/importlib/__init__.py", line 90, in import_module
  File "<frozen importlib._bootstrap>", line 1584, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1565, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1532, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 584, in _check_name_wrapper
  File "<frozen importlib._bootstrap>", line 1022, in load_module
  File "<frozen importlib._bootstrap>", line 1003, in load_module
  File "<frozen importlib._bootstrap>", line 560, in module_for_loader_wrapper
  File "<frozen importlib._bootstrap>", line 868, in _load_module
  File "<frozen importlib._bootstrap>", line 313, in _call_with_frames_removed
  File "C:\Users\xelra\AppData\Roaming\Sublime Text 3\Packages\actualvim\view.py", line 4, in <module>
    from . import neo
  File "C:\Users\xelra\AppData\Roaming\Sublime Text 3\Packages\actualvim\neo.py", line 4, in <module>
    from .lib import neovim
  File "C:\Users\xelra\AppData\Roaming\Sublime Text 3\Packages\actualvim\lib\neovim\__init__.py", line 11, in <module>
    from .msgpack_rpc import (ErrorResponse, child_session, socket_session,
  File "C:\Users\xelra\AppData\Roaming\Sublime Text 3\Packages\actualvim\lib\neovim\msgpack_rpc\__init__.py", line 8, in <module>
    from .event_loop import EventLoop
  File "C:\Users\xelra\AppData\Roaming\Sublime Text 3\Packages\actualvim\lib\neovim\msgpack_rpc\event_loop\__init__.py", line 12, in <module>
    from .asyncio import AsyncioEventLoop
  File "C:\Users\xelra\AppData\Roaming\Sublime Text 3\Packages\actualvim\lib\neovim\msgpack_rpc\event_loop\asyncio.py", line 17, in <module>
    from actualvim.lib import asyncio
  File "C:\Users\xelra\AppData\Roaming\Sublime Text 3\Packages\actualvim\lib\asyncio\__init__.py", line 15, in <module>
    import _overlapped  # Will also be exported.
ImportError: No module named '_overlapped'
```

**Comment by lunixbochs** [comment]
There's some bug with relative imports from asyncio, I worked around it by moving a bunch of stuff around and importing from actualvim.lib... I'll check this out in a bit.

**Comment by xelra** [comment]
I'm subscribing this thread, just give me an update and I'll test again. Thank you for the great plugin.

**Comment by lunixbochs** [comment]
_overlapped is a C module. Maybe someone could build it for the Windows Python 3.3 version embedded in ST3? I'm using asyncio backported from Python 3.4. I also asked Sublime to release a version with a newer Python, which would fix this too, but no fingers crossed for now :)

**Comment by lunixbochs** [comment]
@xelra I included a compiled version of that pyd for win32 and win64 and confirmed ActualVim works on my Windows machine (there's a setting now if it doesn't detect your neovim path, use the command pallete -> ActualVim Settings). Can you try it now?

**Comment by xelra** [comment]
Nice, looking good so far. Really great!

**Comment by xelra** [comment]
Actually it's crashing after some time.
```
Traceback (most recent call last):
  File "C:\Program Files\Sublime Text 3\sublime_plugin.py", line 812, in run_
    return self.run(edit, **args)
  File "C:\Users\Dominik\AppData\Roaming\Sublime Text 3\Packages\actualvim\actual.py", line 39, in run
    v.press(key)
  File "C:\Users\Dominik\AppData\Roaming\Sublime Text 3\Packages\actualvim\view.py", line 247, in press
    self.sync_from_vim()
  File "C:\Users\Dominik\AppData\Roaming\Sublime Text 3\Packages\actualvim\view.py", line 184, in sync_from_vim
    text = '\n'.join(self.buf[:])
  File "C:\Users\Dominik\AppData\Roaming\Sublime Text 3\Packages\actualvim\lib\neovim\api\buffer.py", line 48, in __getitem__
    return self.request('nvim_buf_get_lines', start, end, False)
  File "C:\Users\Dominik\AppData\Roaming\Sublime Text 3\Packages\actualvim\lib\neovim\api\common.py", line 44, in request
    return self._session.request(name, self, *args, **kwargs)
  File "C:\Users\Dominik\AppData\Roaming\Sublime Text 3\Packages\actualvim\lib\neovim\api\nvim.py", line 131, in request
    res = self._session.request(name, *args, **kwargs)
  File "C:\Users\Dominik\AppData\Roaming\Sublime Text 3\Packages\actualvim\lib\neovim\msgpack_rpc\session.py", line 94, in request
    v = self._yielding_request(method, args)
  File "C:\Users\Dominik\AppData\Roaming\Sublime Text 3\Packages\actualvim\lib\neovim\msgpack_rpc\session.py", line 156, in _yielding_request
    return q.get(timeout=1)
  File "./python3.3/queue.py", line 175, in get
queue.Empty
```

**Comment by lunixbochs** [comment]
You're hitting the bug I referenced here:

> It should be fairly stable, but it hangs Sublime if you do some chained commands (like gg, 9j, etc) so don't do them until I figure out wtf is going on there

If you chain keys, the rpc to neovim hangs. I've filed a bug upstream and I'm working around it for now.

**Comment by lunixbochs** [comment]
@xelra can you try again? any other notes on windows support or can I close this?

main remaining bug is self-commands like dd, yy don't work correctly.

**Comment by xelra** [comment]
I've tested again with  https://github.com/lunixbochs/actualvim/commit/ee8e3cf405c72d4fe109d088eb7cc8a60aca9970.

As you can see there are no problems when after loading Subline Text 3 and the plugin. That worked already before though.
```
DPI scale: 1
startup, version: 3126 windows x64 channel: stable
executable: /C/Program Files/Sublime Text 3/sublime_text.exe
working dir: /C/Program Files/Sublime Text 3
packages path: /C/Users/xelra/AppData/Roaming/Sublime Text 3/Packages
state path: /C/Users/xelra/AppData/Roaming/Sublime Text 3/Local
zip path: /C/Program Files/Sublime Text 3/Packages
zip path: /C/Users/xelra/AppData/Roaming/Sublime Text 3/Installed Packages
ignored_packages: ["Vintage"]
pre session restore time: 0.386468
startup time: 0.435468
first paint time: 0.436468
first paint time: 0.437468
reloading plugin Default.auto_indent_tag
reloading plugin Default.block
reloading plugin Default.comment
reloading plugin Default.convert_syntax
reloading plugin Default.copy_path
reloading plugin Default.delete_word
reloading plugin Default.detect_indentation
reloading plugin Default.duplicate_line
reloading plugin Default.echo
reloading plugin Default.exec
reloading plugin Default.fold
reloading plugin Default.font
reloading plugin Default.goto_line
reloading plugin Default.history_list
reloading plugin Default.indentation
reloading plugin Default.install_package_control
reloading plugin Default.kill_ring
reloading plugin Default.mark
reloading plugin Default.new_templates
reloading plugin Default.open_context_url
Unable to fetch update url contents
reloading plugin Default.open_in_browser
reloading plugin Default.pane
reloading plugin Default.paragraph
reloading plugin Default.paste_from_history
reloading plugin Default.profile
reloading plugin Default.quick_panel
reloading plugin Default.run_syntax_tests
reloading plugin Default.save_on_focus_lost
reloading plugin Default.scroll
reloading plugin Default.set_unsaved_view_name
reloading plugin Default.settings
reloading plugin Default.show_scope_name
reloading plugin Default.side_bar
reloading plugin Default.sort
reloading plugin Default.swap_line
reloading plugin Default.switch_file
reloading plugin Default.symbol
reloading plugin Default.transform
reloading plugin Default.transpose
reloading plugin Default.trim_trailing_white_space
reloading plugin CSS.css_completions
reloading plugin Diff.diff
reloading plugin HTML.encode_html_entities
reloading plugin HTML.html_completions
reloading plugin 0_package_control_loader.00-package_control
reloading plugin Package Control.1_reloader
reloading plugin Package Control.2_bootstrap
reloading plugin Package Control.Package Control
reloading plugin actualvim.actual
reloading plugin actualvim.edit
reloading plugin actualvim.neo
reloading plugin actualvim.view
plugins loaded
reloading settings Packages/User/ActualVim.sublime-settings
Package Control: No updated packages
```
Also after using chained commands, I don't see any errors like in the log from my last comment. They also kinda work. For example I was able to successfully do `"*p`.

They're quite buggy though. For example it now deletes lines on `d` instead of `dd`. Also it yanks lines on `y` instead of `yy`. If I do a `yp` it pastes the current line below, when it should do that with `yyp`. So that behavior is quite consistent.

There is also a performance or response issue when holding down a key. For example holding down `j` or `l` doesn't show the cursor moving.

I don't think the above mentioned things are necessarily Windows related though.

**Comment by lunixbochs** [comment]
Oh, you didn't notice I merged neovim to master and have been committing there. ee8e3cf is old. I worked around chained [dyc] chars in a later commit against master. Not sure what's up with holding j. Works for me on OS X if I enable key repeat in Sublime as per the readme. I'll try it Windows (or maybe it works in master. I changed how key buffering works)

**Comment by xelra** [comment]
I've upgraded to https://github.com/lunixbochs/actualvim/commit/2f7d723f9a0f56267def6098d0591797ef4d6848.

The aforementioned problems with chained commands are gone. The only thing that I could find not working is the problem I mentioned with holding down keys. As long as multiple keys are pressed or any key is held down, the cursor doesn't update. Not shift though.

**Comment by lunixbochs** [comment]
> As long as multiple keys are pressed

Can you be more specific about that?

**Comment by xelra** [comment]
For example, when I hold down `a` and then type anything else, the text will only show after releasing all the keys. Or when I type so fast that I'm already pressing another key before fully releasing the other.

**Comment by lunixbochs** [comment]
That doesn't make any sense. Is your key repeat rate very high? What happens when you hold A and type other letters into Sublime Text with actualvim disabled?

I explicitly support updating the UI between keypresses:

https://github.com/lunixbochs/actualvim/blob/master/view.py#L303

Key input works like this:
When you press a key in sublime, it invokes the actualvim keypress text command. This calls press(key) on the instance for the active view. press() puts the pressed key into a queue, then takes a global lock and pops one key from the queue (the queue prevents keys from going out of order if multiple key presses are waiting on the lock at once). When vim is finished handling each key, if vim is "ready" i.e. not halfway into a multi-key command, it syncs cursor and buffer state back to sublime, and only after this releases the lock. As sync happens between key presses, the view/cursor is guaranteed to update after each key if vim is ready, and no keys should ever contend with each other over the view state due to the lock.

I've tested this on OS X with extreme key mashing and had no display speed problems (including logging when multiple keys are waiting at once, which does happen). I'll test it on Windows and see if something is different.

**Comment by xelra** [comment]
With Actual Vim disabled, the behavior is normal: I long press for example `a` and then the first `a` appears on the screen, and after a slight delay many `a`s appear in rapid succession.

With Actual Vim enabled, the behavior is as described: I long press `a` and then nothing appears until I release `a`. Then suddenly all the input appears, in this case `aaaaaaaaaaaaaaaaaaaaaaa`. It looks like everything is working fine in the background, it's just invisible.

**Comment by lunixbochs** [comment]
The UI *completely* blocks on Windows between key presses, it's nothing like I see on OS X.

**Comment by lunixbochs** [comment]
That said, I improved the heuristics and Windows should enjoy much faster input in many cases now (and OS X is as low as 6ms for a keystroke now).

**Comment by xelra** [comment]
I'm on the newest commit. The issue is partly fixed. I can hold down `a` now and produce `aaaaaaaaaaa` and see each `a` updated to the UI. 
It doesn't work with `backspace` though. Holding it down will still not update the UI in between.

**Comment by lunixbochs** [comment]
Okay, I can say with reasonable certainty the freezing is related to how long a keypress takes to process. For me, frozen keypresses take about 90ms, while working ones take around 20ms. This explains why OS X at <20ms was working fine.

The typing-really-fast bug is annoying, it's due to cursor not synchronizing correctly. I've spent a lot of time fixing it, and have ideas on how to make it better if it's still present. It's the reason there are locks taken out everywhere.

**Comment by lunixbochs** [comment]
Backspace and a few other keys are now also considered "simple keys" which bypass the async check.

**Comment by xelra** [comment]
I can reliably reproduce it with medium effort.

When typing fast for a prolonged time, at some point it will start freezing for a bit. When that happens some of the keystrokes are interpreted as commands. After it unfreezes, for example pastes from the clipboard will show up. It's still in insert mode though.

**Comment by lunixbochs** [comment]
Ooh, if it freezes some actions will dump an <esc> into vim to break the loop. That's probably happening, though it's not supposed to on normal typing. Can you put print('force esc') in neo.py in the force_ready function, restart sublime, and see if it's hitting that around the breakage?

**Comment by mn3m** [comment]
Hi @lunixbochs, thanks for the nice work on ActualVim so far! :)

You mentioned bugs regarding dd further up, but anyways: I'm seeing some weird behaviour on Windows. When attempting e.g. 5dd, the results are one deleted line after typing 5d. Then if I type dd after that, five lines are deleted.

**Comment by lunixbochs** [comment]
You're on the master branch right? That bug should only be in the neovim branch.

**Comment by lunixbochs** [comment]
@xelra I found a bug in the insert shortcut that could chain with another bug to escape insert mode when it shouldn't, so I assume there's another bug, but I fixed the first one in 8c96667

**Comment by xelra** [comment]
There seem to have been some improvements. I cannot get any default keys to appear on the right side of the cursor anymore, however, all the non-default keys appear on the right side of the cursor. I'm using a non-English keyboard. The ones that consistently appear on the right side of the cursor are `³²µ€ÄÖÜäö´ß§°`.

**Comment by lunixbochs** [comment]
I don't intercept non-ascii keys yet, see: https://github.com/lunixbochs/actualvim/blob/master/Default.sublime-keymap

**Comment by lunixbochs** [comment]
I think everything here has its own issue, so I'm closing Windows support as fixed.
